gtsfm/utils/metrics.py

```diff
@@ -0,0 +1,179 @@
+from typing import Dict, List, Optional, Tuple, Union
+
+from gtsam import Point3, Rot3, Pose3, Unit3
+
+import gtsfm.utils.geometry_comparisons as comp_utils
+from gtsfm.common.sfm_result import SfmResult
+
+# A StatsDict is a dict from string to optional floats or their lists.
+StatsDict = Dict[str, Union[Optional[float], List[Optional[float]]]]
+
+
+def compute_errors_statistics(errors: List[Optional[float]]) -> StatsDict:
+    """Computes statistics (min, max, median) on the given list of errors
+
+    Args:
+        errors: List of errors for a metric.
+
+    Returns:
+        A dict with keys min_error, max_error, median_error,
+        and errors_list mapping to the respective stats.
+    """
+
+    metrics = {}
+    valid_errors = [error for error in errors if error is not None]
+    metrics["median_error"] = np.median(valid_errors)
+    metrics["min_error"] = np.min(valid_errors)
+    metrics["max_error"] = np.max(valid_errors)
+    metrics["errors_list"] = errors
+    return metrics
+
+
+def compute_rotation_angle_metrics(wRi_list: List[Optional[Rot3]], gt_wRi_list: List[Optional[Pose3]]) -> StatsDict:
+    """Computes statistics for the angle between estimated and GT rotations.
+
+    Assumes that the estimated and GT rotations have been aligned and do not
+    have a gauge freedom.
+
+    Args:
+        wRi_list: List of estimated camera rotations.
+        gt_wRi_list: List of ground truth camera rotations.
+
+    Returns:
+        A statistics dict of the metrics errors in degrees.
+    """
+    errors = []
+    for (wRi, gt_wRi) in zip(wRi_list, gt_wRi_list):
+        errors.append(comp_utils.compute_relative_rotation_angle(wRi, gt_wRi))
+    return compute_errors_statistics(errors)
+
+
+def compute_translation_distance_metrics(
+    wti_list: List[Optional[Point3]], gt_wti_list: List[Optional[Point3]]
+) -> StatsDict:
+    """Computes statistics for the distance between estimated and GT translations.
+
+    Assumes that the estimated and GT translations have been aligned and do not
+    have a gauge freedom (including scale).
+
+    Args:
+        wti_list: List of estimated camera translations.
+        gt_wti_list: List of ground truth camera translations.
+
+    Returns:
+        A statistics dict of the metrics errors in degrees.
+    """
+    errors = []
+    for (wti, gt_wti) in zip(wti_list, gt_wti_list):
+        errors.append(comp_utils.compute_points_distance_l2(wti, gt_wti))
+    return compute_errors_statistics(errors)
+
+
+def compute_translation_angle_metrics(
+    i2Ui1_dict: Dict[Tuple[int, int], Optional[Unit3]], wTi_list: List[Optional[Pose3]]
+) -> StatsDict:
+    """Computes statistics for angle between translations and direction measurements.
+
+    Args:
+        i2Ui1_dict: List of translation direction measurements.
+        wTi_list: List of estimated camera poses.
+
+    Returns:
+        A statistics dict of the metrics errors in degrees.
+    """
+    angles = []
+    for (i1, i2) in i2Ui1_dict:
```

Comment:
i might prefer to explicitly say `i2Ui1_dict.keys()` to emphasize its not a list

Reply:
Style guides generally prefer using default iterators when iterating over lists and dicts. 
https://google.github.io/styleguide/pyguide.html#28-default-iterators-and-operators
The only disadvantage is as you said, it is less explicit that it is a dict, but I think the name of the variable and the type hint should make that clear. Let me know if this is really a concern and I can change it.